09_job_queues/video_processing.py

```diff
@@ -0,0 +1,191 @@
+#     # get the video from the form
+#     video = request.files.get("video")
+#     # get the user_id from the form
+#     user_id = request.form.get("user_id")
+
+#     # generate random project id
+#     project_id = uuid.uuid4()
+
+#     # upload the video to google cloud storage
+#     video_url, folder_name = upload_video_to_gcs(project_id, title, video)
+
+#     # create a project entry in the database
+#     insert_project(project_id, user_id, 0, video_url, title, folder_name)
+
+#     # send the video processing task to the Celery queue
+#     result = transcribe_video_and_extract_screenshots.delay(project_id, video_url, title)
+
+#     return {"project_id": project_id}
+# ```
+
+
+# ```python
+# # tasks.py
+#
+# import os
+# from celery import Celery
+# from celery.utils.log import get_task_logger
+# from video_utilities import transcribe_video_whisper
+# from database_utilities import fetch_project_data, update_project_status, insert_timestamps_and_text
+
+# app = Celery('tasks', broker=os.getenv("CELERY_BROKER_URL"), backend=os.getenv("CELERY_RESULT_BACKEND"))
+# logger = get_task_logger(__name__)
+
+# @app.task
+# def transcribe_video_and_extract_screenshots(project_id, video_url, title):
+
+#     # fetch the project data from supabase
+#     project_data = fetch_project_data(project_id)
+
+#     # check the status of the project
+#     if project_data["task_status"] == 1:
+#         return
+
+#     # transcribe the video
+#     sentences = transcribe_video_whisper(video_url)
+
+#     timestamps_and_text = extract_screenshots(sentences)
+
+#     # insert the timestamps and text into the database
+#     insert_timestamps_and_text(project_id, timestamps_and_text["phrase_texts"], timestamps_and_text["relevant_frames"])
+
+#     # update the status of the project to be "completed"
+#     update_project_status(project_id, 1)
+
+#     return
+# ```
+
+# The Celery queue was working fine, but it still took on the order of ~1-2 minutes to process a 2-3 minute video. I was hoping to be able to speed this up, and decided to turn to Modal.
+# Some addition requirements: I wanted to minimize the amount of refactoring, and wanted to continue to host the Flask app on Render. All I wanted was to be able to offload the video processing work to a job somewhere, and for that work to complete faster.
+
+# ## The Modal application
+
+# To use Modal instead of a Celery queue, we first create a separate file, `modal_queue.py`, and define a Modal Stub and Image.
+# The image is a Debian image with ffmpeg, wget, git, and the python dependencies installed.
+
+# We also import the helper function from `tasks.py`, `transcribe_video_and_extract_screenshots`, which is doing the bulk of the video processing work.
+# Note that we are importing `transcribe_video_extract_screenshots` from outside of the Modal function. This is because the `tasks.py` file exists locally.
+
+import modal
+from tasks import transcribe_video_and_extract_screenshots
```

Comment:
pytest is complaining about this import, how do you want to handle? tried breaking out tasks you have commented out above into a separate file, but that one also seems to refer to some other modules i can't seem to find.

Reply:
Yeah, I think this example was previously supposed to be on the blog, where the code snippets don't actually have to run...

I guess one option is that I can turn `transcribe_video_and_extract_screenshots` into a dummy function, commenting out its body, and just make note of that in the comments above. Issue is that part of what I wanted to show-case was the migration from a non-Modal job queue to a Modal job queue. I also wanted to show how to import different modules.

Are there no other multi-file examples?